File: benchmark/synduce/constraints/bst/sum_gt_by_key.f
Inductive List = elt {Int, Int} | cons {Int, Int, List};
Inductive Map = value {Int, Int} | node {Int, Map, Map};

min = \a: Int. \b: Int. if < a b then a else b;
max = \a: Int. \b: Int. if > a b then a else b;

min_key = fix (
  \f: Map -> Int. \m: Map.
  match m with
    value {k, v} -> k
  | node {a, l, r} -> min (f l) (f r)
  end
);

max_key = fix (
  \f: Map -> Int. \m: Map.
  match m with
    value {k, v} -> k
  | node {a, l, r} -> max (f l) (f r)
  end
);

is_map = fix (
  \f: Map -> Bool. \m: Map.
  match m with
    value {k, v} -> true
  | node {a, l, r} -> and (and (< (max_key l) a) (<= a (min_key r))) (and (f l) (f r))
  end
);

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {k, v, t} -> cons {k, v, f t y}
  | elt {k, v} -> cons {k, v, y}
  end
);

repr = fix (
  \f: Map -> List. \m: Map.
  match m with
    value {k, v} -> elt {k, v}
  | node {k, l, r} -> cat (f l) (f r)
  end
);

@Input key: Int;

spec = fix (
  \f: List -> Int. \l: List.
  match l with
    elt {k, v} -> if > k key then v else 0
  | cons {k, v, t} ->
    + (if > k key then v else 0) (f t)
  end
);

target = fix (
  \f: Map -> Compress Map. \m: Map.
  match m with
    value {k, v} -> value {k, v}
  | node {a, l, r} ->
    if > a key then node {a, f l, f r}
    else node {a, l, f r} /*Avoid the recursion of l*/
  end
);

main = \m: Map. if is_map m then spec (repr (target m)) else 0;

File: ml-benchmark/synduce/constraints/bst/sum_gt_by_key.ml
type lst = Elt of int * int | Cons of int * int * lst
type map = Value of int * int | Node of int * map * map

let min a b = if a < b then a else b
let max a b = if a > b then a else b

let rec min_key m =
  match m with
  | Value (k, v) -> k
  | Node (a, l, r) -> min (min_key l) (min_key r)

let rec max_key m =
  match m with
  | Value (k, v) -> k
  | Node (a, l, r) -> max (max_key l) (max_key r)

let rec is_map m =
  match m with
  | Value (k, v) -> true
  | Node (a, l, r) -> (max_key l) < a && a <= (min_key r) && is_map l && is_map r

let rec cat x y =
  match x with
  | Cons (k, v, t) -> Cons (k, v, cat t y)
  | Elt (k, v) -> Cons (k, v, y)

let rec repr m =
  match m with
  | Value (k, v) -> Elt (k, v)
  | Node (k, l, r) -> cat (repr l) (repr r)

let key = 0

let rec spec l =
  match l with
  | Elt (k, v) -> if k > key then v else 0
  | Cons (k, v, t) -> (if k > key then v else 0) + spec t

let rec target m =
  match m with
  | Value (k, v) -> Value (k, v)
  | Node (a, l, r) ->
      if a > key then Node (a, target l, target r)
      else Node (a, l, target r)

let program m =
  if is_map m then spec (repr (target m)) else 0